libraries/rush-lib/src/logic/PublishUtilities.ts

```diff
@@ -788,21 +788,19 @@ export class PublishUtilities {
       // If the version range exists and has not yet been updated to this version, update it.
       if (requiredVersion.versionSpecifier !== change.newRangeDependency || alwaysUpdate) {
         let changeType: ChangeType | undefined;
-        if (changeType === undefined) {
-          // Propagate hotfix changes to dependencies
-          if (change.changeType === ChangeType.hotfix) {
-            changeType = ChangeType.hotfix;
-          } else {
-            // Either it already satisfies the new version, or doesn't.
-            // If not, the downstream dep needs to be republished.
-            // The downstream dep will also need to be republished if using `workspace:*` as this will publish
-            // as the exact version.
-            changeType =
-              semver.satisfies(change.newVersion!, requiredVersion.versionSpecifier) &&
-              !isWorkspaceWildcardVersion
-                ? ChangeType.dependency
-                : ChangeType.patch;
-          }
+        // Propagate hotfix changes to dependencies
+        if (change.changeType === ChangeType.hotfix) {
+          changeType = ChangeType.hotfix;
+        } else {
+          // Either it already satisfies the new version, or doesn't.
+          // If not, the downstream dep needs to be republished.
+          // The downstream dep will also need to be republished if using `workspace:*` as this will publish
+          // as the exact version.
+          changeType =
+            !isWorkspaceWildcardVersion &&
+            semver.satisfies(change.newVersion!, requiredVersion.versionSpecifier)
```

Comment:
It's surprising that `semver.satisfies()` would have any significant cost.

How many times does this get called in your setup? Roughly, what was the total execution time?  (Is the perf speedup like 30s -> 1s, or more like 800ms -> 200ms?)

Reply:
The time is all spent constructing `Range` and `SemVer` objects.

Reply:
Yikes.  After discussion, we determined that this graph visitor algorithm is `O(n^2)` and David measured 80M calls of the API.  Fixing that will likely give a more dramatic speed improvement, but is a more risky change. 

Let's merge this PR first, and then file an issue to follow up.